### ISSUE TITLE
ANCM forwards WebSocket upgrade request even if WebSocket Protocol is not enabled in IIS

### ISSUE THREAD

**Issue by analogrelay** [aspnet/AspNetCoreModule#135]
**This requires Full IIS, not IIS Express.**

Repro steps:
1. Uninstall WebSockets Protocol from IIS (NOTE: This was the default prior to Windows Server 2016)
2. Attempt to run an ASP.NET WebSockets application (@BrennanConroy has a sample app)
3. Connect the WebSocket

Expected Behavior: The browser fails to upgrade (because IIS does not support WebSockets)
Actual Behavior: The browser believes the upgrade is complete, but the connection is immediately terminated.

The issue appears to be that if the incoming request is an Upgrade, but WebSockets are disabled in IIS, ANCM still forwards the Upgrade to the application. Thus the application thinks it can set up a WebSocket and sends the response headers for the upgrade back (which ANCM forwards back to the client). However, the socket has not been set up and IIS closes the HTTP connection after the response is transmitted.

ANCM should be handling the Upgrade request in this case and sending back a 400 or other failure status code to indicate that the WebSocket cannot be established.

**Comment by jkotalik** [comment]
The issue seems to be here: https://github.com/aspnet/AspNetCoreModule/blob/dev/src/AspNetCore/Src/forwardinghandler.cxx#L1204-L1213
and 
https://github.com/aspnet/AspNetCoreModule/blob/dev/src/AspNetCore/Src/dllmain.cpp#L168-L175
(I don't know if that second check is correct in the full IIS scenario).

**Comment by pan-wang** [comment]
I would treat this as user misconfiguration issue instead of an ANCM bug. ANCM is just a proxy. It is not ANCM's responsibility to check whether websocket was enabled at either side (IIS and backend).

**Comment by analogrelay** [comment]
I would disagree in this case. I don't expect ANCM to know if WebSockets is enabled in the application, but here we're talking about it being disabled **in IIS**. If WebSockets are not supported by the user's IIS configuration, ANCM should not be forwarding them to the client, since the client has no way of knowing that ANCM is incapable of successfully proxying the response.

ANCM is proxying incorrectly because it is proxying a request to the backend despite knowing it cannot proxy the response back correctly.

**Comment by Tratcher** [comment]
When the server receives an upgrade request, the upgrade is optional on the application's part, so I think ANCM should continue to forward those requests to the app. However, it should not allow the 101 response when it knows WebSockets are disabled. ANCM could also strip some headers from the request if websockets are disabled.

IApplicationBuilder.ServerFeatures was designed to light up this kind of runtime server feature information in the app. I wonder if flowing the state of WebSockets at startup would have value. SignalR is going to need that information right?

**Comment by pan-wang** [comment]
I synced up with Andrew offline. We agreed that ANCM should not detect IIS WebSocket setting. ANCM can forward the upgrade request. Once ANCM fails to forward the 101 response, it will return 400 error (instead of 502) back to the client and immediate terminate the connection with backend.

________________________________

**Comment by Tratcher** [comment]
400? The client didn't do anything wrong. Or do you think they should have known better than to send the WebSocket request?

**Comment by analogrelay** [comment]
400 may not be the right status code. The spec isn't clear on what to do here (mostly because it would be specifying behavior for servers not adhering to the spec :)). However, in testing, I found that google.com does a 400 when you try to connect a WebSocket (which it, of course, does not support). Tested via curl:

```
curl www.google.com -H "User-Agent: Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2228.0 Safari/537.36" -H "Connection: upgrade" -H "Upgrade: websocket" -H "Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==" -H "Origin: http://google.com" -H "Host: google.com" -H "Sec-WebSocket-Version: 13" -I -L
```

Having said that, we weren't seeing a 502, we were seeing a successful 101 but the connection would immediately terminate since there is no WebSocket module handing the frames. So I'm not sure what you mean by sending a 400 **instead** of a 502.

**Comment by pan-wang** [comment]
In existing code, ANCM reset the connection with the client and set http status to 502 which will not reach the client but be logged by IISLog.  I am suggesting to change the error code to 400 and return it back to the client instead of reset the connection.

**Comment by shirhatti** [comment]
@anurse @pan-wang From this thread, I'm unsure to parse what the current behavior is. Has this been addressed already?

**Comment by Tratcher** [comment]
This should be addressed by https://github.com/aspnet/IISIntegration/issues/584

**Comment by analogrelay** [comment]
It will be mitigated by fixing that issue but the main ask wouldn't be addressed. The issue is that Kestrel doesn't know that IIS will not be able to proxy the socket so the app doesn't realize that despite a successful upgrade, the client will fail. This is a scenario that the server app must always be prepared for, since downstream proxies (beyond ANCM) may fail to forward the socket. aspnet/IISIntegration#584 is about improving the communication between ANCM and the app to allow us to detect in advance that WebSockets is disabled.

This issue is a little different in that I was asking for ANCM to simply refuse to forward the WebSocket upgrade request when it knows it will not be able to proxy the WebSocket traffic. If the WebSocket module is not enabled (or is enabled but not supported on the OS), simply return a 400 and never forward the upgrade. That would be useful even with the fix for aspnet/IISIntegration#584

**Comment by Tratcher** [comment]
I don't recommend refusing requests at the ANCM layer. That prevents them from being logged at the application layer and prevents the application from providing a custom error. With the added communication we're adding applications should be able to detect the potential upgrade failure in advance and provide their own error.
